doc/developer-guides/sw_design_guidelines.rst

```diff
@@ -0,0 +1,494 @@
+          pr_err("vcpu is offline \r\n");
+          status = -EINVAL;
+  }
+
+  if (status == 0) {
+          vcpu = vcpu_from_vid(vm, vcpu_id);
+          ...
+  }
+
+
+Module Level
+============
+
+Fulfillment to FuSa Standards
+-----------------------------
+
+The fulfillment matrix of error detection and error handling on module level to
+FuSa Standards [IEC_61508-7]_ in ACRN hypervisor is shown in
+:numref:`fulfillment_matrix_module_level` below.
+
+.. table:: Fulfillment to FuSa Standards on Module Level
+   :align: center
+   :widths: auto
+   :name: fulfillment_matrix_module_level
+
+   +-----------------------+---------------+----------+---------------------------------------------------+
+   | **Technique/Measure** | **Ref in      | **SIL3** | **Company Solution**                              |
+   |                       | IEC 61508-7** |          |                                                   |
+   +=======================+===============+==========+===================================================+
+   | Design and coding     | C.2.6         | HR       | This technique will be used for internal functions|
+   | standards             | Table B.1     |          | of the hypervisor. Plan to use data verification, |
+   |                       |               |          | with explicit specification of pre-conditions and |
+   |                       |               |          | post-conditions for functions.                    |
+   +-----------------------+---------------+----------+---------------------------------------------------+
+
+
+Error Handling Methods
+----------------------
+
+The error handling methods used in ACRN hypervisor on module level are shown
+below.
+
+**Crash the system via ASSERT**
+  The hypervisor can crash the system when the below cases occur. It shall
+  only be used for debug and used to check pre-conditions and post-conditions
+  to catch design error.
+
+  This method applies to the following case:
+
+  - Software design errors occur.
+
+
+Rules of Error Detection and Error Handling
+-------------------------------------------
+
+The rules of error detection and error handling on module level is shown in
+:numref:`rules_module_level` below.
+
+.. table:: Rules of Error Detection and Error Handling on Module Level
+   :align: center
+   :widths: auto
+   :name: rules_module_level
+
+   +--------------------+-----------+----------------------------+---------------------------+-------------------------+
+   | **Resource Class** | **Failure | **Error Detection via      | **Error Handling Policy** | **Example**             |
+   |                    | Mode**    | Hypervisor**               |                           |                         |
+   +====================+===========+============================+===========================+=========================+
+   | Internal data of   | N/A       | Partial no.                | The hypervisor shall use  | virtual PCI device      |
```

Comment:
I'm not sure what ``Partial no`` means.  I think it is a ``No``, or maybe "Incomplete", or maybe "Partial".